### PR TITLE
Support OIDC auth backends

### PR DESCRIPTION
**Pull request by sergeytrasko**
Adding support for OIDC auth backends.

The implementation is an extension of JWT auth backend (reusing most of the logics) + adding some required attributes.

Documentation updated with examples and description of new attributes.
Tested with local Vault 1.1.1

### CODE REVIEW

**Comment by sergeytrasko** [comment]
Any feedback on this PR?

File: vault/resource_jwt_auth_backend.go [hunk 1]
```diff
@@ -50,6 +58,19 @@ func jwtAuthBackendResource() *schema.Resource {
 				Description: "The OIDC Discovery URL, without any .well-known component (base path)",
 			},
 
+			"oidc_client_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Client ID uses for OIDC",
```
Comment (roidelapluie) [inline vault/resource_jwt_auth_backend.go]: used*

File: vault/resource_jwt_auth_backend.go [hunk 2]
```diff
@@ -50,6 +58,19 @@ func jwtAuthBackendResource() *schema.Resource {
...
+			},
+
+			"oidc_client_secret": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Sensitive:   true,
+				Description: "Client Secret uses for OIDC",
```
Comment (roidelapluie) [inline vault/resource_jwt_auth_backend.go]: used*

File: vault/resource_jwt_auth_backend.go (hunk 1 above)
Comment (sergeytrasko) [reply vault/resource_jwt_auth_backend.go]: Fixed

File: vault/resource_jwt_auth_backend.go (hunk 2 above)
Comment (sergeytrasko) [reply vault/resource_jwt_auth_backend.go]: Fixed

**Comment by roidelapluie** [comment]
Do you plan to add bound_claims to the roles too?

**Comment by sergeytrasko** [comment]
Our use case does not expect `bound_claims` usage - we are more interested in basic functionality now.
I can try to add it later to this PR or I can also submit another one later.
What do you think would be the better approach?

**Comment by roidelapluie** [comment]
I will need that feature but do not want to force anyone to code it.

**Comment by sergeytrasko** [comment]
Actually, not only `bound_claims` support was missing. I've added it together with `claim_mappings` and `oidc_scopes`

As for `oidc_scopes`, see also https://github.com/hashicorp/vault-plugin-auth-jwt/issues/44

**Comment by roidelapluie** [comment]
awesome!

File: vault/resource_jwt_auth_backend_role.go [hunk 3]
```diff
@@ -98,6 +106,24 @@ func jwtAuthBackendRoleResource() *schema.Resource {
...
 			},
+			"oidc_scopes": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "List of OIDC scopes to be used with an OIDC role. The standard scope \"openid\" is automatically included and need not be specified",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"bound_claims": {
```
Comment (roidelapluie) [inline vault/resource_jwt_auth_backend_role.go]: wondering if that should be plural or singular

Comment (sergeytrasko) [reply vault/resource_jwt_auth_backend_role.go]: Using plurals to be in line with Vault's documentation - https://www.vaultproject.io/api/auth/jwt/index.html

File: vault/resource_jwt_auth_backend.go [hunk 4]
```diff
@@ -29,6 +28,15 @@ func jwtAuthBackendResource() *schema.Resource {
 				ValidateFunc: validateNoTrailingSlash,
 			},
 
+			"type": {
```
[the remark this answers is not in the document]
Comment (sergeytrasko) [reply vault/resource_jwt_auth_backend.go]: Not sure... This `type` belongs to auth backend not to a role.

**Comment by sergeytrasko** [comment]
Fixed compilation issue - it was result of incorrect merge from master.

**Comment by roidelapluie** [comment]
Awesome

**Comment by sergeytrasko** [comment]
@tyrannosaurus-becks Does it seem ready from your side? Or is there something missing still?

**Pull request merged**